{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ea17ee38",
   "metadata": {},
   "source": [
    "# Method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a91a5348",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install activeSVC==4.0.1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f90167b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import time\n",
    "import random\n",
    "import os\n",
    "\n",
    "from sklearn.preprocessing import normalize \n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.decomposition import PCA\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import gridspec\n",
    "\n",
    "from activeSVC import min_complexity, min_acquisition\n",
    "import pandas as pd\n",
    "import pickle\n",
    "import os, psutil\n",
    "import resource\n",
    "\n",
    "import csv\n",
    "from scipy.io import mmread\n",
    "from scipy import io as sio\n",
    "import copy\n",
    "\n",
    "def text_create(path, name, msg):\n",
    "    full_path = path + \"/\" + name + '.pickle'\n",
    "    f=open(full_path,'wb') \n",
    "    pickle.dump(msg,f)\n",
    "    f.close()\n",
    "\n",
    "class TimerError(Exception):\n",
    "     \"\"\"A custom exception used to report errors in use of Timer class\"\"\"\n",
    "\n",
    "class Timer:\n",
    "    def __init__(self):\n",
    "        self._start_time = None\n",
    "\n",
    "    def start(self):\n",
    "        if self._start_time is not None:\n",
    "            raise TimerError(f\"Timer is running. Use .stop() to stop it\")\n",
    "\n",
    "        self._start_time = time.perf_counter()\n",
    "\n",
    "    def stop(self):\n",
    "        if self._start_time is None:\n",
    "            raise TimerError(f\"Timer is not running. Use .start() to start it\")\n",
    "\n",
    "        elapsed_time = time.perf_counter() - self._start_time\n",
    "        self._start_time = None\n",
    "        print(f\"Total run time: {elapsed_time:0.4f} seconds\")\n",
    "        return elapsed_time\n",
    "        \n",
    "\n",
    "def load_genes(genes):\n",
    "    '''\n",
    "    Load the gene names from a file\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    genes : str\n",
    "        Path to a gene file\n",
    "    '''\n",
    "    try:\n",
    "        genes = np.array([row[1].upper() for row in csv.reader(open(genes), delimiter=\"\\t\")]) # 10X format\n",
    "    except:\n",
    "        genes = np.array([row[0].upper() for row in csv.reader(open(genes), delimiter=\"\\t\")]) # base format with one gene name per row\n",
    "    return genes\n",
    "\n",
    "## Load in the data and create a metadata table\n",
    "def load_multiple_samples(samples, barcodes, genefile):\n",
    "    '''\n",
    "    Load the gene names from a set of mtx files\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    samples : dict\n",
    "        dictionary of sample names and path to matrix file\n",
    "    barcodes : dict\n",
    "        dictionary of sample names and path to barcodes file\n",
    "    genefile : str\n",
    "        Path to a gene file\n",
    "    '''\n",
    "\n",
    "    genes = load_genes(genefile)\n",
    "\n",
    "    samplenames = list(samples.keys())\n",
    "    sampleorder = list(np.sort(samplenames))\n",
    "\n",
    "    Mlist = []\n",
    "    meta=pd.DataFrame()\n",
    "    for x in sampleorder: \n",
    "        print('Loading in sample: ' + x)\n",
    "        currM = sio.mmread(samples[x]).tocsc()\n",
    "        currbc = list(pd.read_csv(barcodes[x], header=None)[0])\n",
    "        Mlist.append(currM)\n",
    "        currmeta = pd.DataFrame({'sample':[x]*currM.shape[1]})\n",
    "        currmeta.index = [currbc[i] + '_' + x for i in range(len(currbc))]  \n",
    "        meta = meta.append(currmeta)\n",
    "\n",
    "    M = ss.hstack(Mlist)\n",
    "\n",
    "    return M, genes, meta\n",
    "\n",
    "def load_consolidated_data(matrixfile, metafile, genefile):\n",
    "\n",
    "    '''\n",
    "    Load the gene names from a set of mtx files\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    matrixfile : string\n",
    "        path to matrix file\n",
    "    metafile : string\n",
    "        path to metadata file \n",
    "    genefile : str\n",
    "        Path to a gene file\n",
    "    '''\n",
    "    genes = load_genes(genefile)\n",
    "    meta = pd.read_csv(metafile, header=0) \n",
    "    M = sio.mmread(matrixfile).tocsc()    \n",
    "    return M, genes, meta\n",
    "\n",
    "def normalization(M, meta):\n",
    "    '''\n",
    "    Normalize databy dividing by column sum\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    M : sparse matrix\n",
    "        gene expression matrix\n",
    "    meta : table\n",
    "        metadata table\n",
    "    '''\n",
    "\n",
    "    # normalize by dividing by column sum\n",
    "    M2 = copy.deepcopy(M)\n",
    "    sums = np.array(M2.sum(axis=0)).flatten() # compute sums of all columns (cells)\n",
    "    M2.data = M2.data.astype(float) # convert type from int to float prior to division\n",
    "\n",
    "    for i in range(len(M2.indptr)-1): # for each column i\n",
    "        rr = range(M2.indptr[i], M2.indptr[i+1]) # get range rr\n",
    "        M2.data[rr] = M2.data[rr]/sums[i] # divide data values by matching column sum\n",
    "\n",
    "    # add transcript totals to the metadata table\n",
    "    meta['transcript_total'] = sums\n",
    "\n",
    "    return M2, meta\n",
    "\n",
    "def filter_genes(M, offset):\n",
    "    '''\n",
    "    Find indices of genes which have suprapoisson coefficient of variation\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    M : array\n",
    "        genes-cells matrix\n",
    "    offset : float (between 1 and 1.3)\n",
    "        log10(offset) is added to the regression line when filtering genes. \n",
    "        Genes above the line are kept\n",
    "        Values between 1 (no offset) and 1.3 work well\n",
    "\n",
    "    '''\n",
    "    # select genes above the line \n",
    "    mean, var = mean_variance_axis(M, axis=1) # get genes means and variances from M\n",
    "    std = np.sqrt(var) # compute standard deviations from variances\n",
    "    cv = np.divide(std, mean, out=np.zeros_like(std), where=mean!=0) # compute coefficient of variation\n",
    "\n",
    "    # indices of genes that are present in more than .1% of the cells\n",
    "    MCSR = M.tocsr() # MCSR is the row oriented version of M\n",
    "    presence = np.array([MCSR.indptr[i+1]-MCSR.indptr[i] for i in range(M.shape[0])]) # count how many cells have non zeros expression for each gene\n",
    "    presence_idx = np.where(presence>M.shape[1]*0.001)[0] # get indices of genes that are expressed in more than .1% of the cells\n",
    "    MCSR = None\n",
    "\n",
    "    nzidx = np.nonzero(mean)[0] # indices of genes with non-zero mean\n",
    "    nzidx = np.intersect1d(nzidx, presence_idx) # get intersection of genes with non-zero mean and genes present in more than .1% of the cells\n",
    "\n",
    "    nzcv = cv[nzidx] # select the matching cvs\n",
    "    nzmean = mean[nzidx] # select the matching means\n",
    "    lognzcv = np.log10(nzcv) # log10\n",
    "    lognzmean = np.log10(nzmean) # log10\n",
    "\n",
    "    slope, intercept, r_value, p_value, std_err = linregress(lognzmean, lognzcv)\n",
    "\n",
    "    adjusted_intercept = intercept+np.log10(offset) # slide filtering line with offset\n",
    "    selection_idx = np.where(lognzcv>lognzmean*slope+adjusted_intercept)[0] # get indices of genes above the filtering line\n",
    "    final_idx = nzidx[selection_idx]\n",
    "\n",
    "    # Plot the data just to see the selected genes\n",
    "    plt.figure(figsize=(8,6))\n",
    "    plt.scatter(lognzmean, lognzcv, s=1) # plot all genes\n",
    "    plt.scatter(lognzmean[selection_idx], lognzcv[selection_idx], c='maroon', s=1) # plot genes above line with different color\n",
    "    plt.plot(lognzmean,lognzmean*slope+adjusted_intercept, c='darkorange') # plot filtering line\n",
    "    plt.xlabel('log10(mean)')\n",
    "    plt.ylabel('log10(cv)')\n",
    "    plt.title('%d genes selected' % len(selection_idx))\n",
    "\n",
    "    return final_idx\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef9fd3ef",
   "metadata": {},
   "source": [
    "# Load Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b83c185a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import urllib.request\n",
    "import shutil\n",
    "\n",
    "try:\n",
    "    os.mkdir('data')\n",
    "except OSError:\n",
    "    print (\"Creation of the directory %s failed\" % 'data')\n",
    "else:\n",
    "    print (\"Successfully created the directory %s \" % 'data')\n",
    "    \n",
    "with urllib.request.urlopen(\"https://caltech.box.com/shared/static/zywfv43qpeq0jydq80jgas016a3k7pc1.tsv\") as response, open('./data/genes.tsv', \"wb\") as out_file:\n",
    "    shutil.copyfileobj(response, out_file)\n",
    "\n",
    "with urllib.request.urlopen(\"https://caltech.box.com/shared/static/0hlu0xanenjet6klf4zk74quk4o26oyw.csv\") as response, open('./data/MM_all_metadata.csv', \"wb\") as out_file:\n",
    "    shutil.copyfileobj(response, out_file)\n",
    "\n",
    "with urllib.request.urlopen(\"https://caltech.box.com/shared/static/6iam30ygfutxehoy43xuowwhm77y2nxw.mtx\") as response, open('./data/MM_all_matrix.mtx', \"wb\") as out_file:\n",
    "    shutil.copyfileobj(response, out_file)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3c988cea",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_file='./data/MM_all_matrix.mtx'\n",
    "gene_file='./data/genes.tsv'\n",
    "label_file='./data/MM_all_metadata.csv'\n",
    "\n",
    "M, raw_gene, meta = load_consolidated_data(data_file, label_file, gene_file)\n",
    "M2, meta = normalization(M, meta)\n",
    "meta = pd.read_csv(label_file,header=0)\n",
    "\n",
    "l=[]\n",
    "for i in range(np.shape(raw_gene)[0]):\n",
    "    if raw_gene[i][:3]!='RPL' and raw_gene[i][:3]!='RPS' and raw_gene[i][:2]!='MT':\n",
    "        l.append(i)\n",
    "gene=raw_gene[l]\n",
    "\n",
    "data=normalize(M2.tocsr().transpose()[:,l],axis=1, norm='l2')\n",
    "\n",
    "raw_target=meta[\"disease\"].values\n",
    "target=(raw_target=='MM')*1.0\n",
    "target=target.astype(np.uint8)\n",
    "classes=range(len(np.unique(target)))\n",
    "keys=['healthy','MM']\n",
    "del M,M2,raw_gene,meta,raw_target\n",
    "\n",
    "idx = np.arange(np.shape(data)[0])\n",
    "random.shuffle(idx)\n",
    "X_train = data[idx[:int(np.shape(data)[0]*4/5)],:]\n",
    "y_train = target[idx[:int(np.shape(data)[0]*4/5)]]\n",
    "X_test = data[idx[int(np.shape(data)[0]*4/5):],:]\n",
    "y_test = target[idx[int(np.shape(data)[0]*4/5):]]\n",
    "\n",
    "print(type(data))\n",
    "print(np.shape(data),np.shape(target),len(np.unique(target)))\n",
    "print(np.shape(X_train))\n",
    "print(np.shape(X_test))\n",
    "for i in np.unique(target):\n",
    "    print('class '+keys[i]+': '+str(np.count_nonzero((target==i)*1)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b5fb3069",
   "metadata": {},
   "source": [
    "# Select Genes and Save Results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73fa349a",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "min-complexity random\n",
    "'''\n",
    "num_features = 40\n",
    "num_samples=20\n",
    "init_samples=20\n",
    "balance=False\n",
    "\n",
    "path='results/random_40genes_20cells'\n",
    "\n",
    "try:\n",
    "    os.mkdir('results')\n",
    "except OSError:\n",
    "    print (\"Creation of the directory %s failed\" % 'results')\n",
    "else:\n",
    "    print (\"Successfully created the directory %s \" % 'results')\n",
    "try:\n",
    "    os.mkdir(path)\n",
    "except OSError:\n",
    "    print (\"Creation of the directory %s failed\" % path)\n",
    "else:\n",
    "    print (\"Successfully created the directory %s \" % path)\n",
    "    \n",
    "    \n",
    "if __name__ == '__main__':\n",
    "\n",
    "    t=Timer()\n",
    "    t.start()\n",
    "    feature_selected, num_samples_list, train_errors,test_errors,train_scores,test_scores, step_times= min_complexity(\n",
    "        X_train,y_train,X_test,y_test,num_features=num_features,num_samples=num_samples,init_samples=init_samples, balance=balance)\n",
    "    elapsed_time=t.stop()\n",
    "    \n",
    "    memorys=[]\n",
    "    memorys.append(psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2)\n",
    "    memorys.append(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss/ 1024 ** 2)\n",
    "    \n",
    "    text_create(path,'feature_selected',feature_selected)\n",
    "    text_create(path,'error',train_errors+test_errors)\n",
    "    text_create(path,'accuracy',train_scores+test_scores)\n",
    "    text_create(path,'num_samples_list',num_samples_list)\n",
    "    text_create(path,'genes_name',gene[feature_selected])\n",
    "    text_create(path,'elapsed_time',elapsed_time)\n",
    "    text_create(path,'memory',memorys)\n",
    "\n",
    "    plt.figure(figsize=(8,8))\n",
    "    plt.plot(train_scores,linewidth=2)\n",
    "    plt.plot(test_scores,linewidth=2)\n",
    "    plt.legend(['train acc','test acc'],prop = {'size':18})\n",
    "    plt.xlabel('number of genes',fontdict={'weight':'normal','size': 18})\n",
    "    plt.ylabel('accuracy',fontdict={'weight':'normal','size': 18})\n",
    "    plt.tick_params(labelsize=18)\n",
    "    plt.savefig(path+'/acc.pdf', bbox_inches=\"tight\")\n",
    "\n",
    "feature1=np.copy(feature_selected)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c310a6ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "min-complexity balance\n",
    "'''\n",
    "num_features = 40\n",
    "num_samples=20\n",
    "init_samples=20\n",
    "balance=True\n",
    "\n",
    "path='results/balance_40genes_20cells'\n",
    "\n",
    "try:\n",
    "    os.mkdir('results')\n",
    "except OSError:\n",
    "    print (\"Creation of the directory %s failed\" % 'results')\n",
    "else:\n",
    "    print (\"Successfully created the directory %s \" % 'results')\n",
    "try:\n",
    "    os.mkdir(path)\n",
    "except OSError:\n",
    "    print (\"Creation of the directory %s failed\" % path)\n",
    "else:\n",
    "    print (\"Successfully created the directory %s \" % path)\n",
    "    \n",
    "    \n",
    "if __name__ == '__main__':\n",
    "\n",
    "    t=Timer()\n",
    "    t.start()\n",
    "    feature_selected, num_samples_list, train_errors,test_errors,train_scores,test_scores, step_times= min_complexity(\n",
    "        X_train,y_train,X_test,y_test,num_features=num_features,num_samples=num_samples,init_samples=init_samples, balance=balance, class_weight='balanced')\n",
    "    elapsed_time=t.stop()\n",
    "    \n",
    "    memorys=[]\n",
    "    memorys.append(psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2)\n",
    "    memorys.append(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss/ 1024 ** 2)\n",
    "    \n",
    "    text_create(path,'feature_selected',feature_selected)\n",
    "    text_create(path,'error',train_errors+test_errors)\n",
    "    text_create(path,'accuracy',train_scores+test_scores)\n",
    "    text_create(path,'num_samples_list',num_samples_list)\n",
    "    text_create(path,'genes_name',gene[feature_selected])\n",
    "    text_create(path,'elapsed_time',elapsed_time)\n",
    "    text_create(path,'memory',memorys)\n",
    "\n",
    "    plt.figure(figsize=(8,8))\n",
    "    plt.plot(train_scores,linewidth=2)\n",
    "    plt.plot(test_scores,linewidth=2)\n",
    "    plt.legend(['train acc','test acc'],prop = {'size':18})\n",
    "    plt.xlabel('number of genes',fontdict={'weight':'normal','size': 18})\n",
    "    plt.ylabel('accuracy',fontdict={'weight':'normal','size': 18})\n",
    "    plt.tick_params(labelsize=18)\n",
    "    plt.savefig(path+'/acc.pdf', bbox_inches=\"tight\")\n",
    "\n",
    "feature2=np.copy(feature_selected)    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3926b4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "min-cell\n",
    "'''\n",
    "num_features = 150\n",
    "num_samples=100\n",
    "init_samples=100\n",
    "balance=False\n",
    "\n",
    "path='results/mincell_150genes_100cells'\n",
    "\n",
    "try:\n",
    "    os.mkdir('results')\n",
    "except OSError:\n",
    "    print (\"Creation of the directory %s failed\" % 'results')\n",
    "else:\n",
    "    print (\"Successfully created the directory %s \" % 'results')\n",
    "try:\n",
    "    os.mkdir(path)\n",
    "except OSError:\n",
    "    print (\"Creation of the directory %s failed\" % path)\n",
    "else:\n",
    "    print (\"Successfully created the directory %s \" % path)\n",
    "    \n",
    "    \n",
    "if __name__ == '__main__':\n",
    "\n",
    "    t=Timer()\n",
    "    t.start()\n",
    "    feature_selected, num_samples_list, samples_global, train_errors,test_errors,train_scores,test_scores,step_times= min_acquisition(\n",
    "        X_train,y_train,X_test,y_test,num_features=num_features,num_samples=num_samples, init_samples=init_samples)\n",
    "    elapsed_time=t.stop()\n",
    "    \n",
    "    memorys=[]\n",
    "    memorys.append(psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2)\n",
    "    memorys.append(resource.getrusage(resource.RUSAGE_SELF).ru_maxrss/ 1024 ** 2)\n",
    "    \n",
    "    text_create(path,'feature_selected',feature_selected)\n",
    "    text_create(path,'error',train_errors+test_errors)\n",
    "    text_create(path,'accuracy',train_scores+test_scores)\n",
    "    text_create(path,'num_samples_list',num_samples_list)\n",
    "    text_create(path,'samples_global',samples_global)\n",
    "    text_create(path,'genes_name',gene[feature_selected])\n",
    "    text_create(path,'elapsed_time',elapsed_time)\n",
    "    text_create(path,'memory',memorys)\n",
    "\n",
    "    plt.figure(figsize=(8,8))\n",
    "    plt.plot(train_scores,linewidth=2)\n",
    "    plt.plot(test_scores,linewidth=2)\n",
    "    plt.legend(['train acc','test acc'],prop = {'size':18})\n",
    "    plt.xlabel('number of genes',fontdict={'weight':'normal','size': 18})\n",
    "    plt.ylabel('accuracy',fontdict={'weight':'normal','size': 18})\n",
    "    plt.tick_params(labelsize=18)\n",
    "    plt.savefig(path+'/acc.pdf', bbox_inches=\"tight\")\n",
    "\n",
    "    plt.figure(figsize=(8,5))\n",
    "    plt.plot(num_samples_list,linewidth=5)\n",
    "    plt.xlabel('number of genes',fontdict={'weight':'normal','size': 18})\n",
    "    plt.ylabel('number of cells acquired',fontdict={'weight':'normal','size': 18})\n",
    "    plt.tick_params(labelsize=18)\n",
    "    plt.savefig(path+'/cells.pdf', bbox_inches=\"tight\")\n",
    "    \n",
    "feature3=np.copy(feature_selected) \n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8226524",
   "metadata": {},
   "source": [
    "# Plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53a98e84",
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha = 1000\n",
    "data_show = np.log1p(data*alpha).toarray()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1a7aa44e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tsne of original dataset\n",
    "Y1 = pd.read_csv(label_file)[\"UMAP_1\"].values\n",
    "Y2 = pd.read_csv(label_file)[\"UMAP_2\"].values\n",
    "print(np.shape(Y1),np.shape(Y2))\n",
    "\n",
    "plt.figure(figsize=(12,8))\n",
    "c_cmap = plt.get_cmap('Dark2', len(classes))\n",
    "plt.scatter(Y1,Y2,c=target,s=0.5,cmap=c_cmap)\n",
    "cbar=plt.colorbar(ticks=np.arange(len(classes)))\n",
    "plt.xticks([])\n",
    "plt.yticks([])\n",
    "cbar.ax.set_yticklabels(keys,fontdict={'weight':'normal','size': 28})\n",
    "plt.savefig(path+'/tsne.pdf',bbox_inches=\"tight\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6e860131",
   "metadata": {},
   "outputs": [],
   "source": [
    "# tsne of selected genes\n",
    "plt.figure(figsize=(12,8))\n",
    "data_selected=data_show[:,feature_selected]\n",
    "\n",
    "pca=PCA(n_components=40,\n",
    "       copy=True,\n",
    "       whiten=False,\n",
    "       svd_solver='auto',\n",
    "       tol=0.0,\n",
    "       iterated_power='auto',\n",
    "       random_state=None)\n",
    "pcaproj=pca.fit_transform(data_selected)\n",
    "\n",
    "Y_selected=TSNE(n_components=2).fit_transform(pcaproj)\n",
    "\n",
    "c_cmap = plt.get_cmap('Dark2', len(classes))\n",
    "plt.scatter(Y_selected[:,0],Y_selected[:,1],c=target,s=0.5,cmap=c_cmap)\n",
    "cbar=plt.colorbar(ticks=np.arange(len(classes)))\n",
    "plt.xticks([])\n",
    "plt.yticks([])\n",
    "cbar.ax.set_yticklabels(keys,fontdict={'weight':'normal','size': 28})\n",
    "plt.savefig(path+'/tsne_selected.pdf',bbox_inches=\"tight\")\n",
    "import pickle\n",
    "f=open(path+'/tsne_selected.pickle','wb') \n",
    "pickle.dump(Y_selected,f)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2d9a460c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene expression on tsne projection\n",
    "features=list(set(feature1)|set(feature2)|set(feature3))\n",
    "size=len(features)\n",
    "plt.figure(figsize=(30,3*int(size/10)+3))\n",
    "gs = gridspec.GridSpec(int(size/10)+1, 10, width_ratios=[7,7,7,7,7,7,7,7,7,7])\n",
    "\n",
    "for i in range(size):\n",
    "    f=features[i]\n",
    "    ax0 = plt.subplot(gs[i])\n",
    "    ax0.set_title(gene[f],fontdict={'weight':'normal','size': 18})\n",
    "    plt.scatter(Y1,Y2,c=data_show[:,f],s=0.001)\n",
    "    ax0.set_xticks([])\n",
    "    ax0.set_yticks([])\n",
    "plt.savefig(path+'/gene.png',dpi=300,  bbox_inches=\"tight\")    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c8583aac",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
